bridge/tracker/tracker.go

```diff
@@ -0,0 +1,390 @@
+package tracker
+
+import (
+	"context"
+	"encoding/json"
+	"math/big"
+	"os"
+
+	"github.com/0xPolygon/polygon-edge/types"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/syndtr/goleveldb/leveldb"
+	"github.com/umbracle/go-web3"
+)
+
+const (
+	lastQueriedBlockNumber = "last-block-num" //	db key
+)
+
+//	Tracker represents an event listener that notifies
+//	for each event emitted on the rootchain (ethereum).
+//	Events of interest are defined in rootchain.go
+type Tracker struct {
+	logger hclog.Logger
+
+	// required block confirmations
+	confirmations uint64
+
+	// newHeads subscription channel
+	headerCh chan *ethHeader
+
+	// event channel
+	eventCh chan []byte
+
+	// cancel funcs
+	cancelSubscription, cancelEventTracking context.CancelFunc
+
+	//	rootchain client
+	client *rootchainClient
+
+	//	db for last processed block number
+	db *leveldb.DB
+}
+
+//	NewEventTracker returns a new tracker able to listen for events
+//	at the desired block depth. Events of interest are defined
+//	in rootchain.go
+func NewEventTracker(logger hclog.Logger, confirmations uint64) (*Tracker, error) {
+	tracker := &Tracker{
+		logger:        logger.Named("event_tracker"),
+		confirmations: confirmations,
+		headerCh:      make(chan *ethHeader, 1),
+		eventCh:       make(chan []byte),
+	}
+
+	//	create rootchain client
+	client, err := newRootchainClient(rootchainWS)
+	if err != nil {
+		logger.Error("cannot connect to rootchain", "err", err)
+
+		return nil, err
+	}
+
+	tracker.client = client
+
+	//	load db (last processed block number)
+	db, err := tracker.loadDB()
+	if err != nil {
+		logger.Error("cannot load db", "err", err)
+
+		return nil, err
+	}
+
+	tracker.db = db
```

Comment:
This can be a single func `initDB` or something like that, that throws an error if DB initialization is not successful.

You're calling the local method of the `tracker` struct here and just using that value to set an attribute of the same `tracker` struct

Reply:
[Refactored](https://github.com/0xPolygon/polygon-edge/pull/386/commits/8cff0918ede7d6c018f0d99293dd5fa0514ad40d)